components/ai_chat/content/browser/associated_web_contents_content.cc

```diff
@@ -0,0 +1,368 @@
+  }
+
+  // Fetch zero PDF bytes to just receive the total page count.
+  pdf_helper->GetPdfBytes(
+      /*size_limit=*/0,
+      base::BindOnce(&AssociatedWebContentsContent::OnGetPDFPageCount,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void AssociatedWebContentsContent::OnGetPDFPageCount(
+    FetchPageContentCallback callback,
+    pdf::mojom::PdfListener::GetPdfBytesStatus status,
+    const std::vector<uint8_t>& bytes,
+    uint32_t page_count) {
+  auto* pdf_helper =
+      pdf::PDFDocumentHelper::MaybeGetForWebContents(web_contents());
+  if (status == pdf::mojom::PdfListener::GetPdfBytesStatus::kFailed ||
+      !pdf_helper) {
+    std::move(callback).Run("", false, "");
+    return;
+  }
+
+  // Create a barrier callback that will be called when all pages are received
+  auto barrier_callback = base::BarrierCallback<std::pair<size_t, std::string>>(
+      page_count,
+      base::BindOnce(&AssociatedWebContentsContent::OnAllPDFPagesTextReceived,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+
+  for (size_t i = 0; i < page_count; ++i) {
+    pdf_helper->GetPageText(
+        i, base::BindOnce(
+               [](base::OnceCallback<void(std::pair<size_t, std::string>)>
+                      barrier_callback,
+                  size_t page_index, const std::u16string& page_text) {
+                 // Convert to UTF8 immediately and include page index
+                 std::move(barrier_callback)
+                     .Run(std::make_pair(page_index,
+                                         base::UTF16ToUTF8(page_text)));
+               },
+               barrier_callback, i));
+  }
+}
+
+void AssociatedWebContentsContent::OnAllPDFPagesTextReceived(
+    FetchPageContentCallback callback,
+    const std::vector<std::pair<size_t, std::string>>& page_texts) {
+  // Pre-size vector to hold all texts in order
+  std::vector<std::string> ordered_texts(page_texts.size());
+
+  // Insert texts at their correct indices
+  for (const auto& [index, text] : page_texts) {
+    ordered_texts[index] = std::move(text);
+  }
+
+  std::string content;
+  for (auto it = ordered_texts.begin(); it != ordered_texts.end(); ++it) {
+    content.append(*it);
+    if (std::next(it) != ordered_texts.end()) {
+      content.append("\n");
+    }
+  }
+
+  std::move(callback).Run(std::move(content), false, "");
+}
+#endif  // BUILDFLAG(ENABLE_PDF)
+
+void AssociatedWebContentsContent::GetSearchSummarizerKey(
+    GetSearchSummarizerKeyCallback callback) {
+  if (!IsBraveSearchSERP(web_contents()->GetLastCommittedURL())) {
+    std::move(callback).Run(std::nullopt);
+    return;
+  }
+
+  page_content_fetcher_delegate_->GetSearchSummarizerKey(std::move(callback));
+}
+
+void AssociatedWebContentsContent::GetOpenAIChatButtonNonce(
+    mojom::PageContentExtractor::GetOpenAIChatButtonNonceCallback callback) {
+  page_content_fetcher_delegate_->GetOpenAIChatButtonNonce(std::move(callback));
+}
```

Comment:
Does this make sense on the base class or should it only be on AIChatTabHelper?

Reply:
I did try and move it to the `AIChatTabHelper` but it goes through the `page_content_fetcher_delegate` so its hard to separate at the moment.

It'd definitely be good in a follow up though